### PR TITLE
Added requiresSchemaInWhere option to SQLTemplates.

### PR DESCRIPTION
**Pull request by gordski**
This option allows templates to generate where clauses with the pattern
`<SCHEMA>.<TABLE>.<COLUMN>` in the column specs.

Derby doesn't appear to support `<TABLE>.<COLUMN>` when using
`<SCHEMA>.<TABLE>` in the from unless the default schema is set and matches
the `<SCHEMA>` in the from clause.

#2074

### CODE REVIEW

File: querydsl-sql/src/main/java/com/querydsl/sql/SQLTemplates.java [hunk 1]
```diff
@@ -250,14 +252,15 @@ public SQLTemplates build() {
 
     @Deprecated
     protected SQLTemplates(String quoteStr, char escape, boolean useQuotes) {
-        this(Keywords.DEFAULT, quoteStr, escape, useQuotes);
+        this(Keywords.DEFAULT, quoteStr, escape, useQuotes, false);
     }
 
-    protected SQLTemplates(Set<String> reservedKeywords, String quoteStr, char escape, boolean useQuotes) {
+    protected SQLTemplates(Set<String> reservedKeywords, String quoteStr, char escape, boolean useQuotes, boolean requiresSchemaInWhere) {
```
Comment (Shredder121) [inline querydsl-sql/src/main/java/com/querydsl/sql/SQLTemplates.java]: Add in an additional constructor matching the old one.
For backwards compatibility.

**Comment by gordski** [comment]
I've put in the constructor matching the previous signature. I didn't mark it deprecated, happy to change that if you think it should be.

I've left the non-Derby templates using the longer constructor for clarity.

File: querydsl-sql/src/main/java/com/querydsl/sql/SQLSerializer.java [hunk 2]
```diff
@@ -471,7 +471,19 @@ void serializeForDelete(QueryMetadata metadata, RelationalPath<?> entity) {
...
+
+    void serializeForWhere(QueryMetadata metadata) {
+        boolean requireSchemaInWhere = templates.isRequiresSchemaInWhere();
+        if (requireSchemaInWhere) {
+            dmlWithSchema = true;
+        }
+        append(templates.getWhere()).handle(metadata.getWhere());
+
+        if (requireSchemaInWhere) {
+            dmlWithSchema = false;
```
Comment (Shredder121) [inline querydsl-sql/src/main/java/com/querydsl/sql/SQLSerializer.java]: Would it be safer to just store the old value of `dmlWithSchema`, and restore it here?

Comment (gordski) [inline querydsl-sql/src/main/java/com/querydsl/sql/SQLSerializer.java]: Makes sense. Should I change the other methods to do the same?

`serializeForDelete`, `serializeForMerge`,`serializeForInsert` and `serializeForUpdate`

Comment (Shredder121) [inline querydsl-sql/src/main/java/com/querydsl/sql/SQLSerializer.java]: Ohh I didn't know it was already like that.
I suppose that can be changed in a different PR, which we also can do.

Comment (gordski) [inline querydsl-sql/src/main/java/com/querydsl/sql/SQLSerializer.java]: It looks like that member is only changed in that class. There is a `setDmlWithSchema` but I couldn't find any references.

I'm happy to do the new PR as well if you like.

**Comment by Shredder121** [comment]
I think the only thing we are missing is a test.
Could you add a test, which asserts the `query.toString()` to have the schema in the query?

Somewhere in `SQLSerializerTest` looks fine. Maybe name it `schemaInWhere()`?

File: querydsl-sql/src/main/java/com/querydsl/sql/SQLSerializer.java (hunk 2 above)
Comment (Shredder121) [inline querydsl-sql/src/main/java/com/querydsl/sql/SQLSerializer.java]: If you could, that would be awesome!

**Comment by Shredder121** [comment]
Just a question, have you already tried `DerbyTemplates.builder().printSchema()`?
As for now it doesn't yet append the schema.

Because during serialization, in a few visitors it has to know whether to print the schema, but it would need to check first in what stage it is.

**Comment by gordski** [comment]
Yes the issue I have seen is using the `printSchema`. You get `DELETE schema.table WHERE table.column = "foo"` which Derby rejects unless `schema` is the default schema.

My change still only appends the schema to the entities in the `WHERE` if configured with `printSchema`

**Comment by gordski** [comment]
@Shredder121 It would be good to get this merged. Is there anything else I need to do?

**Comment by Shredder121** [comment]
Looks good!
@querydsl/developers it can be merged

**Comment by gordski** [comment]
Great thanks!
